solr/solrj/src/java/org/apache/solr/common/util/SimpleOrderedMap.java

```diff
@@ -68,24 +74,153 @@ public SimpleOrderedMap(Map.Entry<String, T>[] nameValuePairs) {
+   * contained a mapping for the key, the old value is replaced by the specified value.
    *
-   * @return SimpleOrderedMap containing one key-value pair
+   * @param key key with which the specified value is to be associated value – value to be
+   *     associated with the specified key
+   * @param value to be associated with the specified key
+   * @return the previous value associated with key, or null if there was no mapping for key. (A
+   *     null return can also indicate that the map previously associated null with key)
+   */
+  @Override
+  public T put(String key, T value) {
+    int idx = indexOf(key);
+    if (idx == -1) {
+      add(key, value);
+      return null;
+    } else {
+      T t = get(key);
+      setVal(idx, value);
+      return t;
+    }
+  }
+
+  /**
+   * @see NamedList#remove(String)
+   */
+  @Override
+  public T remove(final Object key) {
+    return super.remove((String) key);
+  }
+
+  /**
+   * Copies all of the mappings from the specified map to this map. These mappings will replace any
+   * mappings that this map had for any of the keys currently in the specified map.
+   *
+   * @param m mappings to be stored in this map
+   * @throws NullPointerException if the specified map is null
+   */
+  @Override
+  public void putAll(final Map<? extends String, ? extends T> m) {
+    m.forEach(this::put);
+  }
+
+  /**
+   * return a {@link Set} of all keys in the map.
+   *
+   * @return {@link Set} of all keys in the map
+   */
+  @Override
+  public Set<String> keySet() {
+    return new InnerMap().keySet();
+  }
+
+  /**
+   * return a {@link Collection} of all values in the map.
+   *
+   * @return {@link Collection} of all values in the map
+   */
+  @Override
+  public Collection<T> values() {
+    return new InnerMap().values();
+  }
+
+  /**
+   * Returns a {@link Set} view of the mappings contained in this map.
+   *
+   * @return {@link Set} view of mappings
+   */
+  @Override
+  public Set<Entry<String, T>> entrySet() {
+
+    return new AbstractSet<>() {
+      @Override
+      public Iterator<Entry<String, T>> iterator() {
+        return SimpleOrderedMap.this.iterator();
+      }
+
+      @Override
+      public int size() {
+        return SimpleOrderedMap.this.size();
+      }
+    };
+  }
+
+  /**
+   * Returns an immutable instance of {@link SimpleOrderedMap} with a single key-value pair.
+   *
+   * @return {@link SimpleOrderedMap} containing one key-value pair
    */
   public static <T> SimpleOrderedMap<T> of(String name, T val) {
-    return new SimpleOrderedMap<>(List.of(name, val));
+    return new SimpleOrderedMap<T>(List.of(name, val));
+  }
+
+  /**
+   * Returns a shared, empty, and immutable instance of {@link SimpleOrderedMap}.
+   *
+   * @return Empty {@link SimpleOrderedMap} (immutable)
+   */
+  public static SimpleOrderedMap<Object> of() {
+    return EMPTY;
+  }
+
+  private class InnerMap extends AbstractMap<String, T> {
```

Comment:
If you follow my feedback above, we won't need this.

Reply:
I better appreciate why this is here.  I don't love the name but I don't recommend something else.  I think it should have a comment explaining why this is here.  I'd say "SimpleOrderedMap can't extend AbstractMap, sadly.  This inner class can, and allows us to implement some of Map's methods via one-liner calls out to this.